src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs

```diff
@@ -0,0 +1,338 @@
+                    static (context, ct) => context.TargetSymbol is INamedTypeSymbol methodSymbol
+                        ? new { Syntax = (InterfaceDeclarationSyntax)context.TargetNode, Symbol = methodSymbol }
+                        : null)
+                .Where(
+                    static modelData => modelData is not null);
+
+            var interfacesWithDiagnostics = attributedMethods.Select(static (data, ct) =>
+            {
+                Diagnostic? diagnostic = GetDiagnosticIfInvalidTypeForGeneration(data.Syntax, data.Symbol);
+                return new { data.Syntax, data.Symbol, Diagnostic = diagnostic };
+            });
+
+            // Split the methods we want to generate and the ones we don't into two separate groups.
+            var interfacesToGenerate = interfacesWithDiagnostics.Where(static data => data.Diagnostic is null);
+            var invalidTypeDiagnostics = interfacesWithDiagnostics.Where(static data => data.Diagnostic is not null);
+
+            context.RegisterSourceOutput(invalidTypeDiagnostics, static (context, invalidMethod) =>
+            {
+                context.ReportDiagnostic(invalidMethod.Diagnostic);
+            });
+
+            var methodsWithDiagnostics = interfacesToGenerate
+                .SelectMany(static (data, ct) =>
+            {
+                ContainingSyntaxContext containingSyntax = new(data.Syntax);
+                Location interfaceLocation = data.Syntax.GetLocation();
+                var methods = ImmutableArray.CreateBuilder<(MethodDeclarationSyntax Syntax, IMethodSymbol Symbol, int Index, Diagnostic? Diagnostic)>();
+                // Start at offset 3 as 0-2 are IUnknown.
+                int methodVtableOffset = 3;
+                foreach (var member in data.Symbol.GetMembers())
+                {
+                    if (member.Kind == SymbolKind.Method && !member.IsStatic)
+                    {
+                        // We only support methods that are defined in the same partial interface definition as the
+                        // [GeneratedComInterface] attribute.
+                        // This restriction not only makes finding the syntax for a given method cheaper,
+                        // but it also enables us to ensure that we can determine vtable method order easily.
+                        Location? locationInAttributeSyntax = null;
+                        foreach (var location in member.Locations)
+                        {
+                            if (location.SourceTree == interfaceLocation.SourceTree
+                                && interfaceLocation.SourceSpan.Contains(location.SourceSpan))
+                            {
+                                locationInAttributeSyntax = location;
+                            }
+                        }
+
+                        if (locationInAttributeSyntax is null)
+                        {
+                            methods.Add((
```

Comment:
Are the extra parens needed here?

Reply:
Yes, we're adding a tuple to a list, so it needs the double parens.